File: android/app/src/main/kotlin/com/example/test_twilio/BasicChatClient.kt
package com.example.test_twilio

import ChatCallbackListener
import ChatStatusListener
import ToastStatusListener
import android.content.Context
import android.util.Log
import com.example.test_twilio.arguments.ConversationArgument
import com.example.test_twilio.arguments.MessageItemArgument
import com.example.test_twilio.arguments.SendMessageArgument
import com.example.test_twilio.arguments.UpdateMessageArgument
import com.example.test_twilio.message.MessageClient
import com.twilio.conversations.*
import java.util.*

class BasicChatClient(private val context: Context)
    : CallbackListener<ConversationsClient>
    , ConversationsClientListener, BasicChatClientCallback
{
    private var accessToken: String? = null
    private var fcmToken: String? = null
    private val conversations = HashMap<String?, ConversationModel>()
    private var listener: ChatCallback? = null
    private var messageClient: MessageClient? = null

    private var conversationsClient: ConversationsClient? = null

    init {
        if (BuildConfig.DEBUG) {
            ConversationsClient.setLogLevel(ConversationsClient.LogLevel.VERBOSE)
        }
    }

    fun getConservations(): HashMap<String?, Any?> {
        val hashMap = hashMapOf<String?, Any?>()
        for (entry in conversations.entries) {
            hashMap[entry.key] = entry.value.toMap()
        }
        return hashMap
    }

    fun clearChannel() {
        conversations.clear()
    }

    fun setMainActivityCallback(listener: ChatCallback) {
        this.listener = listener
    }

    fun setFCMToken(fcmToken: String) {
        this.fcmToken = fcmToken
        if (conversationsClient != null) {
            setupFcmToken()
        }
    }


    private fun setupFcmToken() {
        conversationsClient!!.registerFCMToken(ConversationsClient.FCMToken(fcmToken),
                ToastStatusListener(
                        "Firebase Messaging registration successful",
                        "Firebase Messaging registration not successful"))
    }

    fun unregisterFcmToken() {
        conversationsClient!!.unregisterFCMToken(ConversationsClient.FCMToken(fcmToken),
                ToastStatusListener(
                        "Firebase Messaging unRegistration successful",
                        "Firebase Messaging unRegistration not successful"))
    }

    fun createClient(accessToken: String?, firebaseToken: String?) {
        this.accessToken = accessToken
        this.fcmToken = firebaseToken
        if (conversationsClient != null) {
            shutdown()
        }
        val props = ConversationsClient.Properties.newBuilder()
                .setRegion("us1")
                .setDeferCertificateTrustToPlatform(false)
                .createProperties()

        ConversationsClient.create(context.applicationContext,
                this.accessToken!!,
                props,
                this)
    }

    fun updateAccessToken(accessToken: String?) {
        conversationsClient?.let {
            this.accessToken = accessToken
            it.updateToken(accessToken, ChatStatusListener(
                    success = {
                        Log.e(this@BasicChatClient.javaClass.simpleName, "Update accessToken success")
                    },
                    fail = {
                        Log.e(this@BasicChatClient.javaClass.simpleName, "Update accessToken failed")
                    }
            ))
        }
    }

    private fun shutdown() {
        conversationsClient!!.removeListener(this)
        conversationsClient!!.shutdown()
        conversationsClient = null // Client no longer usable after shutdown()
    }

    // Client created, remember the reference and set up UI
    override fun onSuccess(client: ConversationsClient) {
        conversationsClient = client
        conversationsClient?.addListener(this)

        if (fcmToken != null) {
            setupFcmToken()
        }

//        listener?.onCreateBasicChatClientComplete()
    }

    override fun onConversationAdded(conversation: Conversation?) {
        Log.e(this@BasicChatClient.javaClass.simpleName, "onConversationAdded")
        conversations[conversation?.sid] = ConversationModel(conversation!!)
        refreshChannelList()
    }

    override fun onConversationUpdated(conversation: Conversation?, reason: Conversation.UpdateReason?) {
        Log.e(this@BasicChatClient.javaClass.simpleName, "onConversationUpdated")
        conversations[conversation?.sid] = ConversationModel(conversation!!)
        refreshChannelList()
    }

    override fun onConversationDeleted(conversation: Conversation?) {
        Log.e(this@BasicChatClient.javaClass.simpleName, "onConversationDeleted")
        conversations.remove(conversation?.sid)
        refreshChannelList()
    }

    override fun onConversationSynchronizationChange(conversation: Conversation?) {
        Log.e(this@BasicChatClient.javaClass.simpleName, "onConversationSynchronizationChange")
        refreshChannelList()
    }

    // Client not created, fail
    override fun onError(errorInfo: ErrorInfo?) {
        conversationsClient = null
    }

    // Token expiration events
    override fun onTokenAboutToExpire() {
        if (conversationsClient != null) {
          listener?.generateNewAccessToken()
        }
    }

    override fun onTokenExpired() {
        accessToken = null
        if (conversationsClient != null) {
            listener?.generateNewAccessToken()
        }
    }

    override fun onClientSynchronization(synchronizationStatus: ConversationsClient.SynchronizationStatus?) {
        Log.e(this@BasicChatClient.javaClass.simpleName, "onClientSynchronization: ${synchronizationStatus?.toString()}")
        if (synchronizationStatus == ConversationsClient.SynchronizationStatus.COMPLETED) {
            listener?.onCreateBasicChatClientComplete()
        }
    }
    override fun onConnectionStateChange(connectionState: ConversationsClient.ConnectionState?) {
        Log.e(this@BasicChatClient.javaClass.simpleName, "onConnectionStateChange: ${connectionState.toString()}")
    }
    override fun onUserSubscribed(p0: User?) {}
    override fun onUserUnsubscribed(p0: User?) {}
    override fun onUserUpdated(p0: User?, p1: User.UpdateReason?) {}
    override fun onNewMessageNotification(p0: String?, p1: String?, p2: Long) {}
    override fun onAddedToConversationNotification(conversationSid: String?) {}
    override fun onRemovedFromConversationNotification(conversationSid: String?) {}
    override fun onNotificationFailed(p0: ErrorInfo?) {}
    override fun onNotificationSubscribed() {}

    fun getListConversation() {
        conversationsClient?.myConversations?.forEach { conversation ->
            Log.e(this@BasicChatClient.javaClass.simpleName, "getListConversation: ${conversation.friendlyName}")
            Log.e(this@BasicChatClient.javaClass.simpleName, "getListConversation: ${conversation.sid}")
            conversations[conversation.sid] = ConversationModel(conversation)
        }
        refreshChannelList()
    }

    private fun refreshChannelList() {
        Log.e(this@BasicChatClient.javaClass.simpleName, "refreshChannelList")
        listener?.refreshChannelList()
    }

    fun getMessages(conversationArgument: ConversationArgument) {
        Log.e(this.javaClass.simpleName, "getMessages: $conversationArgument")
        val conversationModel = conversations[conversationArgument.sid]
        conversationModel?.getConversation { conversation ->
            messageClient = MessageClient.getInstance()
            messageClient?.setConversation(conversation)
            messageClient?.setBasicChatClientCallback(this@BasicChatClient)
            messageClient?.addListener()
            messageClient?.getLastMessage()
        }
    }

    fun getMessageItemList(): List<HashMap<String, Any?>> {
        val listHashMap = mutableListOf<HashMap<String, Any?>>()

        messageClient?.messageItemList?.let {
            for (messageItem in it) {
                listHashMap.add(messageItem.toMap(listener))
            }
        }
        return listHashMap
    }

    fun removeConversationListener() {
        messageClient?.removeListener()
    }

    override fun getMessageCompleted() {
        listener?.refreshMessagesList()
    }

    fun sendMessage(sendMessageArgument: SendMessageArgument) {
        messageClient?.sendMessage(sendMessageArgument)
    }

    fun joinConversation(conversationArgument: ConversationArgument) {
        val conversationModel = conversations[conversationArgument.sid]
        conversationModel?.getConversation { conversation ->
            conversation?.join(ChatStatusListener(success = {
                Log.e(this@BasicChatClient.javaClass.simpleName, "join success")
                listener?.joinChannelSuccess()
            }, fail = {
                Log.e(this@BasicChatClient.javaClass.simpleName, "join error")
                listener?.joinChannelError()
            }))
        }
    }

    fun deleteMessage(messageId: String) {
        messageClient?.deleteMessage(messageId)
    }

    override fun removeMessageSuccess(messageItemArgument: MessageItemArgument) {
        listener?.removeMessageSuccess(messageItemArgument)
    }

    fun updateMessage(messageArgument: UpdateMessageArgument) {
        messageClient?.updateMessage(messageArgument.messageId, messageArgument.messageBody)
    }

    override fun updateMessageSuccess(messageItemArgument: MessageItemArgument) {
        listener?.updateMessageSuccess(messageItemArgument)
    }

    fun createConversation(conversationName: String) {
        conversationsClient?.createConversation(conversationName, ChatCallbackListener(
                success = {
                    Log.e(this.javaClass.simpleName, "createChannel success: $it")
                    listener?.createChannelResult(true)
                    conversations[it.sid] = ConversationModel(it)
                    refreshChannelList()
                },
                fail = { errorInfo ->
                    Log.e(this.javaClass.simpleName, "createChannel failed: $errorInfo")
                    listener?.createChannelResult(errorInfo.message)
                }
        ))
    }

    fun inviteByIdentity(identity: String) {
        messageClient?.addParticipantByIdentity(identity)
    }

    fun typing() {
        messageClient?.typing()
    }

    override fun onTypingStarted(description: String) {
        listener?.onTypingStarted(description)
    }

    override fun onTypingEnded(description: String) {
        listener?.onTypingEnded(description)
    }

    override fun loadMoreMessageComplete(list: MutableList<MessageItemArgument>) {
        listener?.loadMoreMessageComplete(list)
    }

    fun getMessageBefore() {
        messageClient?.getMessageBefore()
    }

    fun sendFile(sendMessageArgument: SendMessageArgument) {
        messageClient?.sendFile(sendMessageArgument)
    }

    override fun addMessageSuccess(messageItemArgument: MessageItemArgument) {
        listener?.addMessageSuccess(messageItemArgument)
    }
}

interface BasicChatClientCallback {
    fun getMessageCompleted()
    fun addMessageSuccess(messageItemArgument: MessageItemArgument)
    fun removeMessageSuccess(messageItemArgument: MessageItemArgument)
    fun updateMessageSuccess(messageItemArgument: MessageItemArgument)
    fun onTypingStarted(description: String)
    fun onTypingEnded(description: String)
    fun loadMoreMessageComplete(list: MutableList<MessageItemArgument>)
}

File: ios/Runner/BasicChatClient.swift
//
//  BasicChatClient.swift
//  Runner
//
//  Created by KienNT on 10/26/21.
//

import Foundation
import TwilioConversationsClient

class BasicChatClient: NSObject {
    static let instance = BasicChatClient()
    var conversationsClient: TwilioConversationsClient? = nil
    var accessToken: String? = nil
    var firebaseToken: String? = nil
    private var delegate: AppDelegateCallback? = nil
    private var conversations: [String: ConversationModel] = [:]
    
    func setAppDelegateCallback(delegate: AppDelegateCallback) {
        self.delegate = delegate
    }
    
    func updateAccessToken(accessToken: String) {
        self.accessToken = accessToken
        if let conversationsClient = conversationsClient {
            conversationsClient.updateToken(accessToken) { (result) in
                if (result.isSuccessful) {
                    print("update access token success")
                } else {
                    print("update access token failed")
                }
            }
        }
    }
    
    func getConversations() -> [String: Any?]{
        var map: [String: Any?] = [:]
        conversations.forEach { (arg0) in
            let (key, value) = arg0
            map["\(key)"] = value.toMap()
        }
        return map
    }
    
    func getMessageItemList() -> [[String: Any?]] {
        var list: [[String: Any?]] = []
        MessageClient.instance.messageItemList.forEach { (messageItem) in
            list.append(messageItem.toMap())
        }
        return list
    }
    
    func clearChannel() {
        conversations.removeAll()
    }
    
    func createClient(accessToken: String?, firebaseToken: String?) {
        self.accessToken = accessToken
        self.firebaseToken = firebaseToken
        if (conversationsClient != nil) {
            shutdown()
        }
        
        let props = TwilioConversationsClientProperties.init()
        props.region = "us1"
        props.deferCertificateTrustToPlatform = false
        
        TwilioConversationsClient.conversationsClient(withToken: accessToken!, properties: props, delegate: self, completion: { (result, conversationClient) in
            if (result.isSuccessful) {
                self.conversationsClient = conversationClient
                if (self.conversationsClient != nil) {
                    self.setFcmToken()
                }
                self.conversationsClient?.delegate = self
            } else {
                self.conversationsClient = nil
            }
        })
    }
    
    func shutdown() {
        conversationsClient!.shutdown()
        conversationsClient = nil
    }
    
    private func setFcmToken() {
        //TODO: register firebase token
    }
    
    func getListConversation() {
        conversationsClient?.myConversations()?.forEach({ (conversation) in
            conversations[conversation.sid!] = ConversationModel(conversation: conversation)
        })
        refreshChannelList()
    }
    
    func getmessages(conversationArgument: ConversationArgument) {
        let conversationModel = conversations[conversationArgument.sid!]
        conversationModel?.getConversation(listener: { (conversation) in
            MessageClient.instance.setConversation(conversation: conversation)
            MessageClient.instance.setDelegate(delegate: self)
            MessageClient.instance.addListener()
            MessageClient.instance.getLastMessage()
        })
    }
    
    func removeConversationListener() {
        MessageClient.instance.removeListener()
    }
    
    func joinConversation(conversationArgument: ConversationArgument) {
        let conversationModel = conversations[conversationArgument.sid!]
        conversationModel?.getConversation(listener: { (conversation) in
            conversation.join { (result) in
                if (result.isSuccessful) {
                    self.delegate?.joinChannelSuccess()
                } else {
                    self.delegate?.joinChannelError()
                }
            }
        })
    }
    
    func sendMessage(message: String) {
        MessageClient.instance.sendMessage(message: message)
    }
    
    func deleteMessage(messageId: String) {
        MessageClient.instance.deleteMessage(messageId: messageId)
    }
    
    func updateMessage(messageArgument: UpdateMessageArgument) {
        MessageClient.instance.updateMessage(messageId: messageArgument.messageId, body: messageArgument.messageBody)
    }
    
    func createConversation(conversationName: String) {
        conversationsClient!.createConversation(options: [TCHConversationOptionFriendlyName: conversationName]) { (result, conversation) in
            if (result.isSuccessful) {
                print("create conversation success")
                self.delegate?.createChannelResult(result: true)
                self.conversations[conversation!.sid!] = ConversationModel(conversation: conversation!)
                self.refreshChannelList()
            } else {
                print("create conversation error")
                self.delegate?.createChannelResult(result: result.error!.description)
            }
        }
    }
    
    func inviteByIdentity(identity: String) {
        MessageClient.instance.addParticipantByIdentity(identity: identity)
    }
    
    func typing() {
        MessageClient.instance.typing()
    }
    
    func getMessageBefore() {
        MessageClient.instance.getMessageBefore()
    }
}

extension BasicChatClient: TwilioConversationsClientDelegate {
    func conversationsClientTokenExpired(_ client: TwilioConversationsClient) {
        accessToken = nil
        if (conversationsClient != nil) {
            delegate?.generateNewAccessToken()
        }
    }
    
    func conversationsClientTokenWillExpire(_ client: TwilioConversationsClient) {
        if (conversationsClient != nil) {
            delegate?.generateNewAccessToken()
        }
    }
}

extension BasicChatClient: TCHConversationDelegate {
    func conversationsClient(_ client: TwilioConversationsClient, conversationDeleted conversation: TCHConversation) {
        print("onConversationDeleted: \(conversation)")
        conversations.removeValue(forKey: conversation.sid!)
        refreshChannelList()
    }
    
    func conversationsClient(_ client: TwilioConversationsClient, conversation: TCHConversation, updated: TCHConversationUpdate) {
        print("onConversationUpdated: \(conversation)")
        conversations[conversation.sid!] = ConversationModel(conversation: conversation)
        refreshChannelList()
    }
    
    func conversationsClient(_ client: TwilioConversationsClient, conversationAdded conversation: TCHConversation) {
        print("onConversationAdded: \(conversation)")
        conversations[conversation.sid!] = ConversationModel(conversation: conversation)
        refreshChannelList()
    }
    
    func conversationsClient(_ client: TwilioConversationsClient, synchronizationStatusUpdated status: TCHClientSynchronizationStatus) {
        if (status == TCHClientSynchronizationStatus.completed) {
            delegate?.onCreateBasicChatClientComplete()
        }
    }
    
    func conversationsClient(_ client: TwilioConversationsClient, conversation: TCHConversation, synchronizationStatusUpdated status: TCHConversationSynchronizationStatus) {
        refreshChannelList()
    }

    private func refreshChannelList() {
        delegate?.refreshChannelList()
    }
}

extension BasicChatClient: BasicChatClientDelegate {
    func getMessageCompleted() {
        delegate?.refreshMessagesList()
    }
    
    func removeMessageSuccess(messageItemArgument: MessageItemArgument) {
        delegate?.removeMessageSuccess(messageItemArgument: messageItemArgument)
    }
    
    func updateMessageSuccess(messageItemArgument: MessageItemArgument) {
        delegate?.updateMessageSuccess(messageItemArgument: messageItemArgument)
    }
    
    func onTypingStarted(description: String) {
        delegate?.onTypingStarted(description: description)
    }
    
    func onTypingEnded(description: String) {
        delegate?.onTypingEnded(description: description)
    }
    
    func loadMoreMessageComplete(list: [MessageItemArgument]) {
        delegate?.loadMoreMessageComplete(list: list)
    }
    
    
}

protocol BasicChatClientDelegate {
    func getMessageCompleted()
    func removeMessageSuccess(messageItemArgument: MessageItemArgument)
    func updateMessageSuccess(messageItemArgument: MessageItemArgument)
    func onTypingStarted(description: String)
    func onTypingEnded(description: String)
    func loadMoreMessageComplete(list: [MessageItemArgument])
}
